core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -272,6 +277,77 @@ public int getCapacity()
         return hashCapacity;
     }
 
+    private int putIfAbsent(int firstPosition, int positions, Page page, @Nullable BlockBuilder groupIdsBuilder)
+    {
+        /*
+         * Instead of looking for a match on a per-position basis, we execute steps in a "vectorized" way,
+         * so that all cpu mechanisms (caching, out-of-order execution, etc.) are fully utilized.
+         */
+        long[] rawHashes = new long[positions];
+        int[] hashPositions = new int[positions];
+        long[] groupAddresses = new long[positions];
+
+        for (int i = 0; i < positions; i++) {
+            rawHashes[i] = hashGenerator.hashPosition(firstPosition + i, page);
+        }
+
+        for (int i = 0; i < positions; i++) {
+            hashPositions[i] = (int) getHashPosition(rawHashes[i], mask);
+        }
+
+        for (int i = 0; i < positions; i++) {
+            groupAddresses[i] = groupAddressByHash[hashPositions[i]];
```

Comment:
Curious - did you try running without introducing `groupAddresses` at all and keeping access via `groupAddressByHash` always? It seems we are not saving much here.